src/search/BracketsSearchCursor.js

```diff
@@ -0,0 +1,823 @@
+            getGroupValue: function (groupNumber, valueIndexWithinGroup) {return array[(groupSize * groupNumber) + valueIndexWithinGroup]; },
+            currentGroupIndex: function () { return _currentGroupIndex; },
+            currentGroupNumber: function () { return _currentGroupIndex / groupSize; },
+            groupSize: function () { return groupSize; },
+            itemCount: function () { return array.length / groupSize; },
+
+        });
+        return array;
+    }
+
+    /**
+     * Performs a search using the supplied RegExp and adds all results
+     * of matched locations to the group array
+     *
+     * @private
+     * @param   {object}     query                           regular expression query
+     * @param   {String}     docText                         the text to search
+     * @param   {GroupArray} groupArray                      group array to hold values of index locations
+     * @param   {number}     [matchCountLimit=10000000]      search will stop when match limit met
+     * @param   {number}     [searchEndIndex=docText.length] search will stop when last match exceeds index location
+     * @returns {number}                                     Count of results
+     */
+    function _searchAndAddResultsToArray(query, docText, groupArray, matchCountLimit, searchEndIndex) {
+        var matchArray;
+        var index = 0;
+        searchEndIndex = searchEndIndex || docText.length;
+        matchCountLimit = matchCountLimit || 10000000;
+        matchCountLimit *= groupArray.groupSize();
+
+        while ((matchArray = query.exec(docText)) !== null) {
+            groupArray[index++] = matchArray.index;
+            groupArray[index++] = query.lastIndex;
+            // This is to stop infinite loop.  Some regular expressions can return 0 length match
+            // which will not advance the lastindex property.  Ex ".*"
+            if (matchArray.index === query.lastIndex) {
+                query.lastIndex++;
+            }
+            if ((index >= matchCountLimit) || (query.lastIndex > searchEndIndex)) {
+                break;
+            }
+        }
+
+        return index / groupArray.groupSize();
+    }
+
+    /**
+     * Determines if the values of specified items in a group array are equal
+     * @private
+     * @param   {GroupArray} groupArray1 First group array
+     * @param   {GroupArray} groupNum1   Item within group array
+     * @param   {GroupArray} groupArray2 Second group array
+     * @param   {GroupArray} groupNum2   Item within 2nd group array
+     * @returns {boolean}  true when values are equal
+     */
+    function _isGroupValuesEqual(groupArray1, groupNum1, groupArray2, groupNum2) {
+        if (_.isEqual(groupArray1.getGroupValue(groupNum1, 0), groupArray2.getGroupValue(groupNum2, 0)) &&
+                _.isEqual(groupArray1.getGroupValue(groupNum1, 1), groupArray2.getGroupValue(groupNum2, 1))) {
```

Comment:
Either group size should not be a variable (i.e. always assume 2) or this implementation should be made size-independent (again, easy to do if equals belongs to the type).